mantidproject/mantid
# Automatically Finding the Centre of Rotation on a Sample

- [x] ~Requires #18095 merged first.~

## Description of work.

#### NOTE: This PR should only check the new functionality of `reconstruction_command.py` -> the `find_cor` function, and the functions used within. The rest of the script is currently unstable and will be fixed in #18379, as there have been too many broken things to be covered in this PR as well.

1. The reconstruction scripts -> `tomo_reconstruct.py` and `reconstruction_command.py`:
   - Added additional functions to support finding the center of rotation.
   - The new functionality can be accessed if a `-f` or `--find-cor` are passed.

2. GUI:
   - Auto COR functionality now works. 
      - It requires TomoPy installed. 
      - It can be accessed from `ROI etc.` tab, the `Auto-find` button.
      - It uses the selected `Compute Resource`, i.e. if SCARF is selected a SCARF submission will be attempted.
   - Center of Rotation is automatically read from the `StdOut` of the process
      - The way the Center of Rotation is read is quite -> [hacky](https://github.com/mantidproject/mantid/blob/54a29f24e71fa76d733f15f50643a668c4313d30/MantidQt/CustomInterfaces/src/Tomography/TomographyROIViewQtWidget.cpp#L1332), I'm open to suggestions for improvement. The last output from the find centre of rotation run is expected to _always_ be the number of the Centre of Rotation.
   - Center of Rotation is automatically loaded into the GUI
   - BUG FIX: Prompt for cancelling the reconstruction prompt only activates if the process successfully executed
   - BUG FIX: GUI would crash Mantid if the external process crashed/failed to launch, as it would still try to read the empty string and segfault.
   - Loading an image through the `ROI etc.` tab _now changes_ the paths for the sample in the `Setup` tab
   - Depending on whether a `Reconstruction` or a `Auto COR` has been run, the name of the process is now different in the `Reconstruction Jobs` on the `Run` tab

3. ImageROI is renamed to TomographyROI as I think it is clearer. More suggestions for name changes are welcome.

4. The system tests for the reconstruction scripts -> `ImagingIMATTomoScripts` are completely _disabled_ and will be rewritten in #18386. Unit tests are still present, however they will too be rewritten by the mentioned issue. 

<br>

## To test:

- Open Mantid Tomographic Reconstruction GUI
- Go on Setup tab
- Go on Run tab, select Local as computing resource, as mentioned above, it uses the selected `Compute Resource`, i.e. if SCARF is selected a SCARF submission will be attempted.
- Select a small sample (the one from #18095 will do)
- Flat and Dark paths should be detected automatically
- Go on `ROI etc.` tab and Load an image (or the whole stack)
- Check that the paths in the `Setup` tab are accurate, they should be automatically changed
- Auto-find button for Center of Rotation should now be active with only 1 option -> TomoPy

  - Windows:
    - Try running the auto find (it will likely fail)
    - See that the process was added in the Reconstruction Jobs list
    - Mantid should still run fine after this, nothing should be changed in the `ROI etc` tab (note: when the external process failed to launch mantid would crash, but that should be fixed [c58083](https://github.com/mantidproject/mantid/pull/18258/commits/c5808354f7cff6a035365592bc57c8b1c49162d7)

  - Linux:
	- Configure the Region of Interest
    - Run the auto-find tool
    - Look at the script's output, it has been changed to be very verbose and to print elapsed times on each step
      - The >> Memory is intentionally left there for now for debugging on SCARF
    - When the process is finished, see that the calculated COR is loaded into the GUI. The `+` for the COR should be placed at exactly the middle of the image and at the correct X position

  - Repeat the above step with those 2 configurations:
    1. With ROI that's not the full image
	2. With ROI that is the full image, i.e. nothing is cropped

  - Make sure the COR's are always correct to the crop, if the ROI is the full image then both the (REL for relative) REL to CROP and REL to FULL should be the same number. The code for that is [here](https://github.com/mantidproject/mantid/blob/55ca7b80eb1f219e102584fe1c1c034ac774c9cc/scripts/Imaging/IMAT/tomorec/reconstruction_command.py#L1319). 



Release Notes added to `imaging.rst`

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####


##### Functional Tests #####


- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.
